### PR TITLE
Reimplement fuzzing instrumentation using Bazel transitions.

### PR DESCRIPTION
**Pull request by stefanbucur**
This approach eliminates the need for inlining the instrumentation options in the bazelrc file and simplifies the adoption of the rules.

### CODE REVIEW

File: fuzzing/instrum_opts.bzl [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    string_type = type("string")
+    list_type = type(["list"])
+
+    if type(value) != list_type:
+        fail("%s should be a list, but was %s" % (name, type(value)))
+    if any([type(element) != string_type for element in value]):
+        fail("%s should be a list of strings" % name)
+
+def instrumentation_opts(copts = [], linkopts = []):
+    """Creates a set of instrumentation options.
```
Comment (markus-kusano) [inline fuzzing/instrum_opts.bzl:32]: nit: is there a word other than "set" that can be used here? When I read set I think of `{one, of, these, sets}`, but this isn't really returning a set.

If there's a way to introduce new types in bazel (i.e., the equivalent of `struct InstrumentationOptions { ... }` in C) then it would be easier to write the doc string here as just as `Creates a new InstrumentationOptions`

Maybe just struct? Or """Creates a new instrumentation options value in the format expected by...""" (this might be too wordy)

File: fuzzing/private/instrument.bzl [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    "//fuzzing:instrum_opts.bzl",
+    "base_opts",
+    "fuzzing_engine_opts",
+    "instrumentation_opts",
+    "sanitizer_opts",
+)
+
+def _merge_opts(left_opts, right_opts):
+    return instrumentation_opts(
+        copts = left_opts.copts + right_opts.copts,
```
Comment (markus-kusano) [inline fuzzing/private/instrument.bzl:27]: Double checking: do we need to do any deduplication when doing list concatenation, or is it OK to have the same copt specified multiple times? Ditto below

**Review by markus-kusano — COMMENTED** [review]
only nits

File: .github/workflows/bazel_test.yml [hunk 3]
```diff
@@ -27,9 +27,12 @@ jobs:
...
-      - name: Run common tests
+      - name: Run unit tests
         run: |
-          bazel test --config=clang //... --test_tag_filters=-fuzz-test --build_tests_only
-      - name: Run dedicated tests
+          bazel test //... --test_tag_filters=-fuzz-test --build_tests_only
+      - name: Run simple smoke test
         run: |
           bazel run //examples:empty_fuzz_test_run --config=asan-libfuzzer -- --timeout_secs=5
+      - name: Run advanced smoke test
```
[the remark this answers is not in the document]
Comment (stefanbucur) [reply .github/workflows/bazel_test.yml]: This is a great point. I've replaced this set of steps with a separate "Smoke testing" job and enabled a matrix of configuration to test multiple combinations of config x fuzz target.

On this occasion, I discovered that the RE2 example actually triggers an MSAN error :) I will report this on the RE2 project itself and we can decide if we want to include this simple fuzz target as a RE2 fuzzer for OSS-Fuzz too. CC @inferno-chromium @oliverchang

File: fuzzing/BUILD [hunk 4]
```diff
@@ -12,12 +12,38 @@
...
+)
+
+string_flag(
+    name = "cc_engine_sanitizer",
+    build_setting_default = "none",
+    values = [
+        "none",
+        "asan",
+        "msan",
+        "msan-repro",
```
[the remark this answers is not in the document]
Comment (stefanbucur) [reply fuzzing/BUILD:48]: Good point - this mode actually only makes sense for MSAN, which has "origin tracking" enabled for uninitialized memory as a separate, slower feature, mostly useful when debugging crash reproducers. ASAN does not have this kind of instrumentation.

In any case, this prompted me to add some code comments describing each of these options.

File: fuzzing/instrum_opts.bzl [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+# limitations under the License.
+
+"""Fuzz test instrumentation options.
+
+Each fuzzing engine or sanitizer instrumentation recognized by the
+//fuzzing:cc_engine_instrumentation and //fuzzing:cc_engine_sanitizer
+configuration flag should be defined here.
+"""
+
+def _check_is_string_list(name, value):
```
[the remark this answers is not in the document]
Comment (stefanbucur) [reply fuzzing/instrum_opts.bzl]: > just a suggestion since the current version is fine too: I think "list_of_strings" reads more naturally than "string_list".

Agreed this can be spelled differently, but I tend to favor "string list" since it is consistent with Starlark naming. For example, [`attr.string_list`](https://docs.bazel.build/versions/master/skylark/lib/attr.html#string_list).

> Also, maybe use "die" or "fail" instead of "check"? E.g., "is_list_of_strings_or_fail". I think that makes it more obvious what the side effects are of this function

Going with your other suggestion to `fail()` in the top-level call.

File: fuzzing/instrum_opts.bzl [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    string_type = type("string")
```
[the remark this answers is not in the document]
Comment (stefanbucur) [reply fuzzing/instrum_opts.bzl]: > Is this typical bazel style to introduce new names (string_type) for type("string")?

I'm not actually sure. I remember seeing this style in some other Starlark file, but I do agree that `type(value) == type(constant)` checks are more readable. They also seem to be the form provided in the [`type()` documentation](https://docs.bazel.build/versions/master/skylark/lib/globals.html#type), so going with that here, too.

File: fuzzing/instrum_opts.bzl (hunk 1 above)
Comment (stefanbucur) [reply fuzzing/instrum_opts.bzl]: I went with "struct". Unfortunately, Starlark doesn't support custom types, so "structs" are the best choice for structured data.

File: fuzzing/instrum_opts.bzl [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+def instrumentation_opts(copts = [], linkopts = []):
+    """Creates a set of instrumentation options.
+
+    Args:
+      copts: A list of compilation options to pass as `--copt`
+        configuration flags.
+      linkopts: A list of linker options to pass as `--linkopt`
+        configuration flags.
+    """
+    _check_is_string_list("copts", copts)
```
[the remark this answers is not in the document]
Comment (stefanbucur) [reply fuzzing/instrum_opts.bzl]: These are all fair points, and I like the fact that we could simplify the interface of the type checking function. I went with creating just a boolean function checking the type of the arguments, and failing inside the top-level API call.

File: fuzzing/private/instrument.bzl (hunk 2 above)
Comment (stefanbucur) [reply fuzzing/private/instrument.bzl:27]: We do not, and I think idempotent options are automatically folded. However, their order matters, e.g., an `-fsanitize=` option can be followed by an `-fno-sanitize=` option that can undo the effect of the former. So I clarified in the docstring below that the `extra_*` attributes are appended to the computed set of instrumentation options.

**Review by stefanbucur — COMMENTED** [review]
Thanks for the thorough review! PTAL.

File: .github/workflows/bazel_test.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (markus-kusano) [reply .github/workflows/bazel_test.yml]: LGTM, feel free to resolve, leaving open for other's to take a look.

File: fuzzing/BUILD (hunk 4 above)
[the remark this answers is not in the document]
Comment (markus-kusano) [reply fuzzing/BUILD:48]: Oh cool, I didn't even know that existed. Code comments sound good to clarify this. Would it also make sense to clarify the name itself? Maybe `msan-origin-tracking` is a bit clearer. Not a big deal though, up to you. Feel free to resolve.

File: fuzzing/private/instrument.bzl (hunk 2 above)
Comment (markus-kusano) [reply fuzzing/private/instrument.bzl:27]: SGTM, thanks.

File: fuzzing/instrum_opts.bzl (hunk 5 above)
[the remark this answers is not in the document]
Comment (markus-kusano) [reply fuzzing/instrum_opts.bzl]: Thanks for pointing out that Starlark already using string_list, SGTM

File: fuzzing/instrum_opts.bzl (hunk 6 above)
[the remark this answers is not in the document]
Comment (markus-kusano) [reply fuzzing/instrum_opts.bzl]: LGTM

File: fuzzing/instrum_opts.bzl (hunk 7 above)
[the remark this answers is not in the document]
Comment (markus-kusano) [reply fuzzing/instrum_opts.bzl]: Thanks, LGTM

**Review by markus-kusano — APPROVED** [review]

File: fuzzing/BUILD (hunk 4 above)
[the remark this answers is not in the document]
Comment (stefanbucur) [reply fuzzing/BUILD]: > Would it also make sense to clarify the name itself? Maybe `msan-origin-tracking` is a bit clearer.

Good point - done. My original worry was that the name would be too long to spell out on the command line, but we're anyway using `--config` values, which can keep the original "-repro" suffix.
